pkgs/stdenv/generic/check-meta.nix

```diff
@@ -437,6 +445,52 @@ let
       && (if config.checkMetaRecursively or false
       then lib.all (d: d.meta.available or true) references
       else true);
+    } // {
+
+      # Each attrset meta.{build,host,target} is an availability
+      # (see lib.meta.checkAvailability for specification) which
+      # must match against the {buildPlatform, hostPlatform,
+      # targetPlatform}.  If any of these does not match, then the
+      # derivation is guaranteed not to build and therefore should
+      # not even be attempted.
+      availability = let meta = attrs.meta or {}; in {
+        build = {
+          bad = internStringifiable (
+            lib.concatMap (pkg: pkg.meta.availability.host.bad or []) referencesOnBuild ++
+            lib.concatMap (pkg: pkg.meta.availability.build.bad or []) referencesOnHost ++
+            meta.badBuildPlatforms or []
+          );
+        };
+        host = {
+          bad = internStringifiable (
+            lib.concatMap (pkg: pkg.meta.availability.target.bad or []) referencesOnBuild ++
+            lib.concatMap (pkg: pkg.meta.availability.host.bad or []) referencesOnHost ++
+            meta.badPlatforms or []
+          );
+
+          # "host.only" is needed to represent the (deprecated,
+          # legacy) meta.platforms attribute, which should be
+          # avoided in favor of meta.hydraPlatforms and
+          # meta.badPlatforms, since in order to set meta.platforms
+          # correctly you need to be able to see into the future and
+          # know what platforms nixpkgs might someday support.
+          # There are no build.only or target.only attrsets because
+          # we (thankfully) don't have meta.platforms equivalents
+          # for those.
+          only = internStringifiable (
+            (if meta?platforms && meta.platforms != lib.platforms.all
+             then [ meta.platforms ]
+             else [ ])
+            ++ lib.concatMap (pkg: pkg.meta.availability.host.only or []) referencesOnHost
+          );
+        };
+        target = {
+          bad = internStringifiable (
+            lib.concatMap (pkg: pkg.meta.availability.target.bad or []) referencesOnHost ++
+            meta.badTargetPlatforms or []
+          );
+        };
```

Comment:
@Artturin, I would appreciate your feedback on whether or not I got the phase offsets right here. I think I did but I'm not super-confident of that.